crates/ironrdp-client/src/app.rs

```diff
@@ -0,0 +1,328 @@
+#![allow(clippy::print_stderr, clippy::print_stdout)] // allowed in this module only
+
+use std::num::NonZeroU32;
+use std::sync::Arc;
+
+use raw_window_handle::{DisplayHandle, HasDisplayHandle};
+use tokio::sync::mpsc;
+use winit::application::ApplicationHandler;
+use winit::dpi::LogicalPosition;
+use winit::event::{self, WindowEvent};
+use winit::event_loop::{ActiveEventLoop, EventLoop};
+use winit::keyboard::ModifiersKeyState;
+use winit::platform::scancode::PhysicalKeyExtScancode;
+use winit::window::{Window, WindowAttributes};
+
+use crate::rdp::{RdpInputEvent, RdpOutputEvent};
+
+type WindowSurface = (Arc<Window>, softbuffer::Surface<DisplayHandle<'static>, Arc<Window>>);
+
+pub struct App {
+    input_event_sender: mpsc::UnboundedSender<RdpInputEvent>,
+    context: softbuffer::Context<DisplayHandle<'static>>,
+    window: Option<WindowSurface>,
+    buffer_size: (u16, u16),
+    input_database: ironrdp::input::Database,
+}
+
+impl App {
+    pub fn new(
+        event_loop: &EventLoop<RdpOutputEvent>,
+        input_event_sender: &mpsc::UnboundedSender<RdpInputEvent>,
+    ) -> anyhow::Result<Self> {
+        // SAFETY: We drop the softbuffer context right before the event loop is stopped, thus making this safe.
+        // FIXME: This is not a sufficient proof and the API is actually unsound as-is.
+        let display_handle = unsafe {
+            std::mem::transmute::<DisplayHandle<'_>, DisplayHandle<'static>>(event_loop.display_handle().unwrap())
+        };
+        let context = softbuffer::Context::new(display_handle)
+            .map_err(|e| anyhow::anyhow!("unable to initialize softbuffer context: {e}"))?;
```

Comment:
issue: We can’t actually guarantee that the event_loop is dropped after the context at this point. It’s only possible to know that at the callsite. Thus the API is unsound and this unsafe block may cause an undefined behavior in the future if we misuse it. I don’t consider this a blocker since this is not part of the public API of the IronRDP crates, and we can address that in the future.